File: csharp/twelve-days/TwelveDays.cs
using System;
using System.Linq;

public static class TwelveDays
{
    private static string Start(string day) => $"On the {day} day of Christmas my true love gave to me, ";

    private static readonly (string day, string present)[] lines = new[]
    {
        ("first", "a Partridge in a Pear Tree"),
        ("second", "two Turtle Doves"),
        ("third", "three French Hens"),
        ("fourth", "four Calling Birds"),
        ("fifth", "five Gold Rings"),
        ("sixth", "six Geese-a-Laying"),
        ("seventh", "seven Swans-a-Swimming"),
        ("eighth", "eight Maids-a-Milking"),
        ("ninth", "nine Ladies Dancing"),
        ("tenth", "ten Lords-a-Leaping"),
        ("eleventh", "eleven Pipers Piping"),
        ("twelfth", "twelve Drummers Drumming"),
    };

    public static string Recite(int verseNumber)
    {
        if(verseNumber < 1 || verseNumber > 12)
            throw new ArgumentException($"{nameof(verseNumber)} should be in the range 1 to 12");
        
        string joiner((string day, string _) verse) => 
            verse.day == "first" ? "." : $", " + (verse.day == "second" ? "and " : "");

        return Enumerable.Range(1, verseNumber).Reverse()
            .Select(i => lines[i - 1])
            .Aggregate(Start(lines[verseNumber - 1].day), (result, verse) => 
                result + $"{verse.present}{joiner(verse)}");
    }

    public static string Recite(int startVerse, int endVerse)
        => string.Join('\n', Enumerable.Range(startVerse, endVerse - startVerse + 1)
            .Select(i => Recite(i)));
}

File: fsharp/twelve-days/TwelveDays.fs
module TwelveDays

let recite start stop = 
    let days = ["first";"second";"third";"fourth";"fifth";"sixth";"seventh";"eighth";"ninth";"tenth";"eleventh";"twelfth"]
    let gifts = [
        "a Partridge in a Pear Tree";"two Turtle Doves";"three French Hens";
        "four Calling Birds";"five Gold Rings";"six Geese-a-Laying";
        "seven Swans-a-Swimming";"eight Maids-a-Milking";"nine Ladies Dancing";
        "ten Lords-a-Leaping";"eleven Pipers Piping";"twelve Drummers Drumming"
    ]
    let line day = 
        let phrases = [day - 1..-1..0] |> Seq.map (fun x -> if x = 0 && day <> 1 then "and " + gifts.[x] else gifts.[x])
        sprintf "On the %s day of Christmas my true love gave to me, %s." days.[day - 1] (System.String.Join (", ", phrases))
    [start..stop] |> List.map line